From ba602c06e7d1f77e0a9b5b8f0458a7258363bc29 Mon Sep 17 00:00:00 2001
From: Bo-Cun Chen <bc-bocun.chen@mediatek.com>
Date: Tue, 29 Apr 2025 22:15:57 +0800
Subject: [PATCH] net: ethernet: mtk_wed: add WDMA disable flow to WiFi L1 SER

Without this patch, the WED driver cannot prevent packets from entering
WDMA during WiFi L1 SER, causing incomplete packets to get stuck in the
PSE. Then results in the PSE buffer management transitioning to a chaotic
state.

Signed-off-by: Bo-Cun Chen <bc-bocun.chen@mediatek.com>
---
 drivers/net/ethernet/mediatek/mtk_eth_soc.h | 12 ++++++++++++
 drivers/net/ethernet/mediatek/mtk_wed.c     |  3 +++
 2 files changed, 15 insertions(+)

diff --git a/drivers/net/ethernet/mediatek/mtk_eth_soc.h b/drivers/net/ethernet/mediatek/mtk_eth_soc.h
index cc2f0c5..b022e71 100644
--- a/drivers/net/ethernet/mediatek/mtk_eth_soc.h
+++ b/drivers/net/ethernet/mediatek/mtk_eth_soc.h
@@ -1901,6 +1901,18 @@ static inline void mtk_ppe_drop_config(struct mtk_eth *eth, u32 config)
 		mtk_w32(eth, config, PSE_PPE_DROP(2));
 }
 
+static inline void mtk_pse_wdma_enable(struct mtk_eth *eth, int id, bool enable)
+{
+	u32 val;
+
+	val = mtk_r32(eth, MTK_FE_GLO_CFG(PSE_WDMA_PORT(id)));
+	if (enable)
+		val &= ~MTK_FE_LINK_DOWN_P(PSE_WDMA_PORT(id));
+	else
+		val |= MTK_FE_LINK_DOWN_P(PSE_WDMA_PORT(id));
+	mtk_w32(eth, val, MTK_FE_GLO_CFG(PSE_WDMA_PORT(id)));
+}
+
 int mtk_gmac_sgmii_path_setup(struct mtk_eth *eth, int mac_id);
 int mtk_gmac_2p5gphy_path_setup(struct mtk_eth *eth, int mac_id);
 int mtk_gmac_gephy_path_setup(struct mtk_eth *eth, int mac_id);
diff --git a/drivers/net/ethernet/mediatek/mtk_wed.c b/drivers/net/ethernet/mediatek/mtk_wed.c
index 6b01d4b..c71a6bb 100644
--- a/drivers/net/ethernet/mediatek/mtk_wed.c
+++ b/drivers/net/ethernet/mediatek/mtk_wed.c
@@ -2072,6 +2072,8 @@ mtk_wed_reset_dma(struct mtk_wed_device *dev)
 	u32 val;
 	int i;
 
+	mtk_pse_wdma_enable(dev->hw->eth, dev->wdma_idx, false);
+
 	for (i = 0; i < ARRAY_SIZE(dev->tx_ring); i++) {
 		if (!dev->tx_ring[i].desc)
 			continue;
@@ -2756,6 +2758,7 @@ mtk_wed_start(struct mtk_wed_device *dev, u32 irq_mask)
 	mtk_wed_amsdu_init(dev);
 
 	mtk_wed_dma_enable(dev);
+	mtk_pse_wdma_enable(dev->hw->eth, dev->wdma_idx, true);
 	dev->running = true;
 }
 
-- 
2.45.2

